posthog/session_recordings/ai/utils.py

```diff
@@ -0,0 +1,206 @@
+import dataclasses
+from datetime import datetime
+
+from typing import List, Dict, Any
+
+from posthog.models.element import chain_to_elements
+
+
+@dataclasses.dataclass
+class SessionSummaryPromptData:
+    # we may allow customisation of columns included in the future,
+    # and we alter the columns present as we process the data
+    # so want to stay as loose as possible here
+    columns: List[str] = dataclasses.field(default_factory=list)
+    results: List[List[Any]] = dataclasses.field(default_factory=list)
+    # in order to reduce the number of tokens in the prompt
+    # we replace URLs with a placeholder and then pass this mapping of placeholder to URL into the prompt
+    url_mapping: Dict[str, str] = dataclasses.field(default_factory=dict)
+
+    def is_empty(self) -> bool:
+        return not self.columns or not self.results
+
+    def column_index(self, column: str) -> int | None:
+        for i, c in enumerate(self.columns):
+            if c == column:
+                return i
+        return None
+
+
+def reduce_elements_chain(session_events: SessionSummaryPromptData) -> SessionSummaryPromptData:
+    if session_events.is_empty():
+        return session_events
+
+    # find elements_chain column index
+    elements_chain_index = session_events.column_index("elements_chain")
+
+    reduced_results = []
+    for result in session_events.results:
+        if elements_chain_index is None:
+            reduced_results.append(result)
+            continue
+
+        elements_chain: str | None = result[elements_chain_index]
+        if not elements_chain:
+            reduced_results.append(result)
+            continue
+
+        # the elements chain has lots of information that we don't need
+        elements = [e for e in chain_to_elements(elements_chain) if e.tag_name in e.USEFUL_ELEMENTS]
+
+        result_list = list(result)
+        result_list[elements_chain_index] = [{"tag": e.tag_name, "text": e.text, "href": e.href} for e in elements]
```

Comment:
TODO: we can use the url mapping/reduction here too to save on tokens

Reply:
I did consider that but wasn't sure if it was possible. Assuming embeddings are generated in different runs we'd have no way of making sure that the reduced values were equivalent across runs